
## Smallseq data analysis pipeline
This repository provides scripts for analysis of data generated by single-cell small RNA sequencing protocol (Faridani et all, 2016)


### Prerequisities
Install cutadapt, cgat (from https://github.com/CGATOxford/cgat), UMI-tools (https://github.com/CGATOxford/UMI-tools), dr_tools.py and argparse.py (https://github.com/danielramskold/S3_species-specific_sequencing), GenomeFetch.py (https://github.com/cgreer/ResearchScripts/blob/master/finalSpliceGraph/GenomeFetch.py)


## Remove UMI from reads: 
Before mapping reads to the reference genome, remove the UMI sequences from the reads and append it to the read name
```
source ~/cgat-install/conda-install/bin/activate cgat-scripts
python src/remove_umi_from_fastq_v4.py -d rawdata_in -o rawdata_out -u NNNNNNNN -p 10
```

## Adapter trimming with cutadapt, and also trims CA from the beginning of the read (with the parameter -u 2)
```
cutadapt -a file:adapters/cutadapt_3prime.fa -e 0.1 -O 1 -u 2 --minimum-length 18 -o out1.fastq -p out2.fastq in1.fastq in2.fastq
```

## Mapping reads to the reference human genome (hg38) with the following STAR options
```
./STAR   --runThreadN 10   --genomeDir path_to_hg38   --readFilesIn in.fq      --readFilesCommand -      --outSAMstrandField intronMotif   --outFilterScoreMin
OverLread 0   --outFilterMatchNmin 18   --outFilterMatchNminOverLread 0   --outFilterMismatchNoverLmax 0.04   --alignIntronMax 1
```

## Remove the clipped reads - i.e reads mapped via soft-clipping.  
```
-Aligned bam files should be available in a format star_hg38.orig/sample1/sample1.bam
python src/remove_softclipped_reads_parallelized_v2.py -i star_hg38.orig -o star_hg38.clipped -p 10 -n 0 -t 3
```

## Extract reads with minimum 40 nt read length (for details about cut-off see the methods section of Faridani et all, 2016)
```
python src/subsample_bam_by_readlen.py -d star_hg38.orig -o star_hg38 -c 40 -m max
```

## Remove PCR dublicates - aka creating absolute molecule counts
```
source ~/cgat-install/conda-install/bin/activate cgat-scripts
python src/remove_umi_dedup.py -d star_hg38_max40 -o star_hg38 -p $nprocs
```

## Merge unique and multi mapped reads into single bam file and index it
```
ls star_hg38 | xargs -I%  echo 'samtools merge star_hg38/'%/%'_uniq+multi.bam star_hg38/'%/%'_unique.bam star_hg38/'%/%'_multi.bam' > mergebams.sh
cat mergebams.sh | parallel -P 80 -n 1
ls star_hg38/*/*uniq+multi.bam | xargs -P 80 -n 1 samtools index
```

## Remove precursor reads
```
python src/remove_reads_with_genomic_TGG.py -i star_hg38_molc_in -o star_hg38_molc -p 10
```

## count molc
```
python src/count_smallrnas.py -i star_hg38_molc -o counts_molc -p 80
python src/merge_countsmallrnaoutput.py counts_molc_merged.txt counts_molc/*/counts_molc.txt
```

## Collapse multi miRNAs only, keep the rest of table intact
```
mv counts_molc.txt counts_molc_preCollapse.txt
python src/collapse_mirnas_on_counts_molc.py -i counts_molc_preCollapse.txt -o counts_molc.txt
```

```
The counts_molc.txt is a table containing molecule counts of all the RNA biotypes in the annotation table
```

## Citation

Please refer to the following article: Faridani et all, 2016



